src/hooks/useTodos.ts

```diff
@@ -0,0 +1,69 @@
+import { useEffect, useState } from 'react';
+import { TodoType } from '../types/Todo';
+import { deleteTodo, getTodos, postTodo } from '../api/todos';
+
+export const useTodos = (userId: number) => {
+  const [todos, setTodos] = useState<TodoType[]>([]);
+  const [tempTodo, setTempTodo] = useState<TodoType | null>(null);
+  const [error, setError] = useState('');
+
+  const showError = (errorMsg: string) => {
+    setError(errorMsg);
+
+    setTimeout(() => setError(''), 3000);
+  };
+
+  const handleAddTodo = async (title: string) => {
+    if (!title) {
+      showError("Title can't be empty");
+
+      return;
+    }
+
+    const body = {
+      title,
+      userId,
+      completed: false,
+    };
+
+    setTempTodo({
+      ...body,
+      id: 0,
+    });
+
+    postTodo(userId, body)
+      .then(todo => setTodos((prevState) => [...prevState, todo]))
+      .catch(() => showError('Unable to add a todo'))
+      .finally(() => setTempTodo(null));
+  };
+
+  const handleDeleteTodo = async (todoId: number) => {
+    deleteTodo(todoId)
+      .then(() => setTodos(
+        prevState => prevState.filter((todo) => todo.id !== todoId),
+      ))
+      .catch(() => showError('Unable to delete a todo'));
+  };
+
+  const handleClearCompleted = (completedTodos: TodoType[]) => {
+    completedTodos.forEach(({ id }) => handleDeleteTodo(id));
```

Comment:
This generally may be problematic if more than one requests fails, as for each on of them you will set and show the error. You could await all these promises and if one of them fails just once show the error

![image](https://github.com/mate-academy/react_todo-app-add-and-delete/assets/87986955/40028e16-f435-49be-a165-faf93bf97c94)

I didn't memoize these functions but like mentioned in the previous comment it could be done so for all of them, remember to add appropriate dependencies to their dependency arrays